deepinv/models/tgv.py

```diff
@@ -163,70 +168,60 @@ def forward(self, y, ths=None):
 
         return self.x2
 
-
-def nabla(I):
-    b, c, h, w = I.shape
-    G = torch.zeros((b, c, h, w, 2), device=I.device).type(I.dtype)
-    G[:, :, :-1, :, 0] = G[:, :, :-1, :, 0] - I[:, :, :-1]
-    G[:, :, :-1, :, 0] = G[:, :, :-1, :, 0] + I[:, :, 1:]
-    G[:, :, :, :-1, 1] = G[:, :, :, :-1, 1] - I[..., :-1]
-    G[:, :, :, :-1, 1] = G[:, :, :, :-1, 1] + I[..., 1:]
-    return G
-
-
-def nablaT(G):
-    b, c, h, w = G.shape[:-1]
-    I = torch.zeros((b, c, h, w), device=G.device).type(
-        G.dtype
-    )  # note that we just reversed left and right sides of each line to obtain the transposed operator
-    I[:, :, :-1] = I[:, :, :-1] - G[:, :, :-1, :, 0]
-    I[:, :, 1:] = I[:, :, 1:] + G[:, :, :-1, :, 0]
-    I[..., :-1] = I[..., :-1] - G[..., :-1, 1]
-    I[..., 1:] = I[..., 1:] + G[..., :-1, 1]
-    return I
-
-
-# # ADJOINTNESS TEST
-# u = torch.randn((4, 3, 100,100)).type(torch.DoubleTensor)
-# Au = nabla(u)
-# v = torch.randn(*Au.shape).type(Au.dtype)
-# Atv = nablaT(v)
-# e = v.flatten()@Au.flatten()-Atv.flatten()@u.flatten()
-# print('Adjointness test (should be small): ', e)
-
-
-def epsilon(I):  # Simplified
-    b, c, h, w, _ = I.shape
-    G = torch.zeros((b, c, h, w, 4), device=I.device).type(I.dtype)
-    G[:, :, 1:, :, 0] = G[:, :, 1:, :, 0] - I[:, :, :-1, :, 0]  # xdy
-    G[..., 0] = G[..., 0] + I[..., 0]
-    G[..., 1:, 1] = G[..., 1:, 1] - I[..., :-1, 0]  # xdx
-    G[..., 1:, 1] = G[..., 1:, 1] + I[..., 1:, 0]
-    G[..., 1:, 2] = G[..., 1:, 2] - I[..., :-1, 1]  # xdx
-    G[..., 2] = G[..., 2] + I[..., 1]
-    G[:, :, :-1, :, 3] = G[:, :, :-1, :, 3] - I[:, :, :-1, :, 1]  # xdy
-    G[:, :, :-1, :, 3] = G[:, :, :-1, :, 3] + I[:, :, 1:, :, 1]
-    return G
-
-
-def epsilonT(G):
-    b, c, h, w, _ = G.shape
-    I = torch.zeros((b, c, h, w, 2), device=G.device).type(G.dtype)
-    I[:, :, :-1, :, 0] = I[:, :, :-1, :, 0] - G[:, :, 1:, :, 0]
-    I[..., 0] = I[..., 0] + G[..., 0]
-    I[..., :-1, 0] = I[..., :-1, 0] - G[..., 1:, 1]
-    I[..., 1:, 0] = I[..., 1:, 0] + G[..., 1:, 1]
-    I[..., :-1, 1] = I[..., :-1, 1] - G[..., 1:, 2]
-    I[..., 1] = I[..., 1] + G[..., 2]
-    I[:, :, :-1, :, 1] = I[:, :, :-1, :, 1] - G[:, :, :-1, :, 3]
-    I[:, :, 1:, :, 1] = I[:, :, 1:, :, 1] + G[:, :, :-1, :, 3]
-    return I
-
-
-# # ADJOINTNESS TEST
-# u = torch.randn((2, 3,100,100,2)).type(torch.DoubleTensor)
-# Au = epsilon(u)
-# v = torch.randn(*Au.shape).type(Au.dtype)
-# Atv = epsilonT(v)
-# e = v.flatten()@Au.flatten()-Atv.flatten()@u.flatten()
-# print('Adjointness test (should be small): ', e)
+    def nabla(self, x):
```

Comment:
nabla seems to be already defined in TV, I would avoid duplicating code (just import it)